File: src/NodeParser/IntrinsicNodeParser.ts
import ts from "typescript";
import { LogicError } from "../Error/Errors.js";
import type { Context } from "../NodeParser.js";
import type { SubNodeParser } from "../SubNodeParser.js";
import type { BaseType } from "../Type/BaseType.js";
import { LiteralType } from "../Type/LiteralType.js";
import { UnionType } from "../Type/UnionType.js";
import { extractLiterals } from "../Utils/extractLiterals.js";
import { isExtendsType } from "../Utils/isExtendsType.js";
import { IntrinsicType } from "../Type/IntrinsicType.js";
import { StringType } from "../Type/StringType.js";

export const intrinsicMethods = {
    Uppercase: (v) => v.toUpperCase(),
    Lowercase: (v) => v.toLowerCase(),
    Capitalize: (v) => v[0].toUpperCase() + v.slice(1),
    Uncapitalize: (v) => v[0].toLowerCase() + v.slice(1),
} as const satisfies Record<string, ((v: string) => string) | undefined>;

function isIntrinsicMethod(methodName: string): methodName is keyof typeof intrinsicMethods {
    return methodName in intrinsicMethods;
}

export class IntrinsicNodeParser implements SubNodeParser {
    public supportsNode(node: ts.KeywordTypeNode): boolean {
        return node.kind === ts.SyntaxKind.IntrinsicKeyword;
    }
    public createType(node: ts.KeywordTypeNode, context: Context): BaseType {
        const methodName = getParentName(node);
        if (!isIntrinsicMethod(methodName)) {
            throw new LogicError(node, `Unknown intrinsic method: ${methodName}`);
        }

        const method = intrinsicMethods[methodName];
        const argument = context.getArguments()[0];

        try {
            const literals = extractLiterals(argument)
                .map(method)
                .map((literal) => new LiteralType(literal));

            if (literals.length === 1) {
                return literals[0];
            }

            return new UnionType(literals);
        } catch (error) {
            if (isExtendsType(context.getReference())) {
                return new IntrinsicType(method, argument);
            }

            return new StringType();
        }
    }
}

function getParentName(node: ts.KeywordTypeNode): string {
    const parent = node.parent;

    if (!ts.isTypeAliasDeclaration(parent)) {
        throw new LogicError(node, "Only intrinsics part of a TypeAliasDeclaration are supported.");
    }

    return parent.name.text;
}

File: internal/parser/intrinsic_node_parser.go
package parser

import (
	"fmt"
	"unicode/utf8"

	"golang.org/x/text/cases"
	"golang.org/x/text/language"

	"github.com/microsoft/typescript-go/shim/ast"

	"github.com/vega/ts-json-schema-generator/internal/types"
)

// intrinsicMethods maps the intrinsic string manipulation type names of
// lib.es5.d.ts to their implementations, mirroring JavaScript's
// String.prototype.toUpperCase/toLowerCase semantics
// (src/NodeParser/IntrinsicNodeParser.ts).
var intrinsicMethods = map[string]func(string) string{
	"Uppercase":    jsToUpperCase,
	"Lowercase":    jsToLowerCase,
	"Capitalize":   func(v string) string { return jsChangeFirstCase(v, jsToUpperCase) },
	"Uncapitalize": func(v string) string { return jsChangeFirstCase(v, jsToLowerCase) },
}

// IntrinsicNodeParser handles the `intrinsic` keyword used by
// Uppercase/Lowercase/Capitalize/Uncapitalize in lib.es5.d.ts, keyed off the
// name of the enclosing type alias (src/NodeParser/IntrinsicNodeParser.ts).
type IntrinsicNodeParser struct{}

func NewIntrinsicNodeParser() *IntrinsicNodeParser {
	return &IntrinsicNodeParser{}
}

func (p *IntrinsicNodeParser) SupportsNode(node *ast.Node) bool {
	return node.Kind == ast.KindIntrinsicKeyword
}

func (p *IntrinsicNodeParser) CreateType(node *ast.Node, ctx *Context, _ *types.ReferenceType) types.Type {
	methodName := getIntrinsicParentName(node)
	method, ok := intrinsicMethods[methodName]
	if !ok {
		panic(fmt.Errorf("unknown intrinsic method: %s", methodName))
	}

	var argument types.Type
	if args := ctx.Arguments(); len(args) > 0 {
		argument = args[0]
	}
	values := types.ExtractLiterals(argument)
	literals := make([]types.Type, len(values))
	for i, value := range values {
		literals[i] = &types.LiteralType{Value: method(value)}
	}
	if len(literals) == 1 {
		return literals[0]
	}
	return types.NewUnionType(literals)
}

func getIntrinsicParentName(node *ast.Node) string {
	parent := node.Parent
	if parent == nil || !ast.IsTypeAliasDeclaration(parent) {
		panic(fmt.Errorf("only intrinsics part of a TypeAliasDeclaration are supported"))
	}
	return parent.AsTypeAliasDeclaration().Name().Text()
}

var (
	upperCaser = cases.Upper(language.Und)
	lowerCaser = cases.Lower(language.Und)
)

// jsToUpperCase mirrors String.prototype.toUpperCase (Unicode default full
// case conversion).
func jsToUpperCase(v string) string { return upperCaser.String(v) }

// jsToLowerCase mirrors String.prototype.toLowerCase.
func jsToLowerCase(v string) string { return lowerCaser.String(v) }

// jsChangeFirstCase mirrors `v[0].toUpperCase() + v.slice(1)`: only the first
// UTF-16 code unit is case-mapped. Panics on empty strings, like the
// TypeScript implementation.
func jsChangeFirstCase(v string, mapCase func(string) string) string {
	if v == "" {
		panic(fmt.Errorf("cannot capitalize an empty string"))
	}
	first, size := utf8.DecodeRuneInString(v)
	if first > 0xFFFF {
		// In JavaScript v[0] is a lone surrogate, which case-mapping leaves
		// unchanged, so the string comes back as-is.
		return v
	}
	return mapCase(string(first)) + v[size:]
}
